### PR TITLE
implements `engine_preparePayload` and `engine_getPayload` (#4346)

### PR DESCRIPTION
**Pull request by tbenr**
## Changelog

- `engine_preparePayload` client implementation and call integrated in block proposal duty preparation
- `engine_getPayload` client implementation and call integrated in block proposal duty production
- `consensus_assembleBlock` removed

### CODE REVIEW

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ExecutionPayloadUtil.java [hunk 1]
```diff
@@ -16,13 +16,22 @@
...
+import tech.pegasys.teku.infrastructure.collections.cache.LRUCache;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.spec.datastructures.execution.ExecutionPayload;
 import tech.pegasys.teku.spec.executionengine.ExecutionEngineChannel;
+import tech.pegasys.teku.ssz.type.Bytes20;
 
 public class ExecutionPayloadUtil {
+  // we are going to store latest payloadId returned by preparePayload for a given slot
+  // in this way validator doesn't need to know anything about payloadId
+  private LRUCache<UInt64, UInt64> slotToPayloadIdMap;
```
Comment (Nashatyrev) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ExecutionPayloadUtil.java]: 1. I would not make spec logic instances stateful 
2. I think it is not safe to address payload solely by slot due to possible reorgs. E.g. validator may ask for payload for empty slot `N` with parent hash for block `N-1`, then receive a (late) block `N` and ask again for payload for slot `N` with parent hash for block `N`

Comment (Nashatyrev) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ExecutionPayloadUtil.java]: I would manage `payloadId`s on the validator duty side

**Review by Nashatyrev — COMMENTED** [review]
Basically fine with just exception of slot->payload cache

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ExecutionPayloadUtil.java (hunk 1 above)
Comment (tbenr) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/ExecutionPayloadUtil.java]: > 1. I would not make spec logic instances stateful
> 2. I think it is not safe to address payload solely by slot due to possible reorgs. E.g. validator may ask for payload for empty slot `N` with parent hash for block `N-1`, then receive a (late) block `N` and ask again for payload for slot `N` with parent hash for block `N`
> I would manage payloadIds on the validator duty side

In my initial experiments on this topic I was implementing the logic on the beacon side but then I was ending up by changing the validator APIs (you need to pass the payloadId somehow) and doing that I was braking the standard interface. Discussing with @ajsutton we ended up agreeing on the quick solution to store it on the beacon side.

About reorgs, I expect the duty preparation to be called again in such cases, ending up by recalling `engine_preparePayload` for the same slot which then produce (eventually) a new `payloadId` that will be set in the cache overwriting the old one.  `engine_getPayload` will be then always using the latest `payloadId` for a given `slot`. This is at least the intended behavior.

**Comment by tbenr** [comment]
@Nashatyrev 
As per offline discussions, moved `slotToPayloadIdMap` to `BlockFactory`.

I Also fixed a bounce of presumed bugs while determining parentBlockHash in preparePayload and getPayload (always in `BlockFactory`)

I added a parentBlock hash check too.

**Review by Nashatyrev — APPROVED** [review]
LGTM

**Pull request merged**